### PR TITLE
Lsm9ds0 v3

### PR DESCRIPTION
**Pull request by mbelluzzo**
Diff from v2:
- a single vector
- added alias
- doc states that operation read/write is of a single byte
- add a #define for the constant used during the conversion

### CODE REVIEW

**Comment by mbelluzzo** [comment]
Merging by the end of the day if there are no comments left.

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+};
+
+static inline void
+_gyro_init_queue(struct lsm9ds0_gyro_data *mdata, uint8_t scale)
+{
+    struct sol_i2c_op *op;
+
+    sol_vector_init(&mdata->queue, sizeof(struct sol_i2c_op));
+    op = sol_vector_append_n(&mdata->queue, 5);
+
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: check for NULL ?

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+static inline void
+_gyro_read_queue(struct lsm9ds0_gyro_data *mdata)
+{
+    struct sol_i2c_op *op;
+
+    sol_vector_init(&mdata->queue, sizeof(struct sol_i2c_op));
+    op = sol_vector_append_n(&mdata->queue, 6);
+
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: check for NULL?

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+_read_gyro_done(void *cb_data, ssize_t status)
+{
+    struct sol_direction_vector gyro;
+    struct lsm9ds0_gyro_data *mdata = cb_data;
+
+    mdata->pending = NULL;
+
+    if (status <= 0) {
+        SOL_ERR("Couldn't read LSM9DS0 Gyroscope.");
+        return;
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: emit error packet?

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+_init_gyro_done(void *data, ssize_t status)
+{
+    struct lsm9ds0_gyro_data *mdata = data;
+
+    mdata->pending = NULL;
+
+    if (status <= 0) {
+        mdata->init = false;
+        SOL_ERR("Couldn't initialize LSM9DS0 Gyroscope.");
+        return;
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: emit error packet?

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+    }
+
+    _gyro_init_queue(mdata, scale_reg);
+
+    mdata->pending = sol_i2c_dispatcher_add_op_set(mdata->i2c, mdata->addr, &mdata->queue,
+        _init_gyro_done, mdata, 0);
+
+    if (!mdata->pending) {
+        SOL_ERR("Couldn't initialize LSM9DS0 Gyroscope.");
+        return -1;
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: You should close the i2c in case this fails

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+static int
+lsm9ds0_gyro_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    struct lsm9ds0_gyro_data *mdata = data;
+
+    if (!mdata->init || mdata->pending)
+        return -1;
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: I guess we should use a better errno value in here

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+static inline void
+_xmt_read_queue(struct lsm9ds0_xmt_data *mdata)
+{
+    struct sol_i2c_op *op;
+
+    sol_vector_init(&mdata->queue, sizeof(struct sol_i2c_op));
+    op = sol_vector_append_n(&mdata->queue, 14);
+    // Accel X Y Z
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: Check for null

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+};
+
+static inline void
+_xmt_init_queue(struct lsm9ds0_xmt_data *mdata, uint8_t accel_reg, uint8_t mag_reg)
+{
+    struct sol_i2c_op *op;
+
+    sol_vector_init(&mdata->queue, sizeof(struct sol_i2c_op));
+    op = sol_vector_append_n(&mdata->queue, 6);
+
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: check for null

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+    uint32_t t;
+    struct sol_i2c_op *t_op;
+    struct sol_direction_vector xmt;
+    struct lsm9ds0_xmt_data *mdata = cb_data;
+
+    mdata->pending = NULL;
+
+    if (status <= 0) {
+        SOL_ERR("Couldn't read LSM9DS0 XMT device.");
+        return;
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: error packet?

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+static void
+_init_xmt_done(void *data, ssize_t status)
+{
+    struct lsm9ds0_xmt_data *mdata = data;
+
+    mdata->pending = NULL;
+
+    if (status <= 0) {
+        mdata->init = false;
+        SOL_ERR("Couldn't initialize LSM9DS0 XMT device.");
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: error packet

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+        mdata->mag_scale = 4;
+    }
+    _xmt_init_queue(mdata, accel_reg, mag_reg);
+
+    mdata->pending = sol_i2c_dispatcher_add_op_set(mdata->i2c, mdata->addr, &mdata->queue,
+        _init_xmt_done, mdata, 0);
+
+    if (!mdata->pending) {
+        SOL_ERR("Couldn't initialize LSM9DS0 XMT device.");
+        return -1;
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: close i2c

File: src/modules/flow/lsm9ds0/lsm9ds0.c [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+static int
+lsm9ds0_xmt_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    struct lsm9ds0_xmt_data *mdata = data;
+
+    if (!mdata->init || mdata->pending)
+        return -1;
```
Comment (cabelitos) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: better errno value

File: src/lib/io/sol-i2c-common.c [hunk 13]
```diff
@@ -68,6 +307,12 @@ sol_i2c_open(uint8_t bus, enum sol_i2c_speed speed)
     i2c_shared->i2c = i2c;
     i2c_shared->refcount = 1;
 
+    i2c_shared->dispatcher.timer = NULL;
+    i2c_shared->dispatcher.pending = NULL;
+    i2c_shared->dispatcher.set_idx = 0;
+    i2c_shared->dispatcher.retry = 0;
```
Comment (cabelitos) [inline src/lib/io/sol-i2c-common.c]: sol_vector_append() returns a 0'ed memory you don't need to do that

File: src/lib/io/sol-i2c-common.c [hunk 14]
```diff
@@ -23,20 +23,259 @@
...
+    if (!i) {
+        // current set is in execution, delay the deletion until current operation finishes
+        op_set->delete_me = true;
+    } else {
+        sol_ptr_vector_del(pv, i);
+        free(op_set);
+    }
+}
+
+void
```
Comment (cabelitos) [inline src/lib/io/sol-i2c-common.c]: missing static

Comment (mbelluzzo) [inline src/lib/io/sol-i2c-common.c]: :+1:

File: src/lib/io/sol-i2c-common.c (hunk 13 above)
Comment (mbelluzzo) [inline src/lib/io/sol-i2c-common.c]: :+1:

File: src/modules/flow/lsm9ds0/lsm9ds0.c (hunk 2 above)
Comment (mbelluzzo) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: :+1:

File: src/modules/flow/lsm9ds0/lsm9ds0.c (hunk 6 above)
Comment (mbelluzzo) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: -EAGAIN will be since isn't like a error, is more like the node isn't read to process stuff yet.

**Comment by mbelluzzo** [comment]
Thanks for the review @cabelitos, merged with the proposed changes!

File: src/modules/flow/lsm9ds0/lsm9ds0.c (hunk 6 above)
Comment (barbieri) [inline src/modules/flow/lsm9ds0/lsm9ds0.c]: the port process errors will be just sent to stderr (or ignored) by the flow engine. If that should have some meaning, then send an error packet.

**Pull request closed without merging**